synapse/handlers/sync.py

```diff
@@ -982,6 +982,10 @@ def _get_rooms_changed(self, sync_result_builder, ignored_users):
             # we do send down the room, and with full state, where necessary
             old_state_ids = None
             if room_id in joined_room_ids or has_join:
+                # Always include if the user (re)joined the room
+                if room_id in joined_room_ids and non_joins:
+                    newly_joined_rooms.append(room_id)
+
```

Comment:
we should probably `continue` here to avoid getting dups in `newly_joined_rooms`